core/src/main/scala/scala/pickling/PicklingErrors.scala

```diff
@@ -0,0 +1,165 @@
+package scala.pickling
+
+private[pickling] object Feedback {
+
+  def failedGenerationMsg(x: String, y: String, z: String) =
+    s"Failed generation of a pickler/unpickler for $x, cannot find a $z for $y."
+
+  def failedParsingMsg(x: String, format: String) =
+    s"Failed to parse $x as $format."
+
+  def disabledRuntimeGenerationMsg(tpe: String, failed: String) =
+    s"Couldn't create $failed for $tpe because runtime generation is disabled."
+
+}
+
+object PicklingErrors {
+
+  import Feedback._
+
+  /** Super class of any pickling exception that can be thrown. */
+  class BasePicklingException(val message: String,
+    val cause: Option[Throwable] = None) extends RuntimeException(message, cause.orNull)
+
+  /** Provides `apply` and `extractors` that allow the new change in
+    * the errors api while keeping the same public old interface.
+    */
+  object BasePicklingException {
+
+    def apply(msg: String, cause: Option[Throwable] = None) =
+      new BasePicklingException(msg, cause)
+
+    def unapply(b: BasePicklingException): Option[(String, Option[Throwable])] =
+      Some(b.message -> b.cause)
+
+  }
+
+  /** Exception thrown when the generation of a [[Pickler]] for a type
+    * fails because scala pickling wasn't able to find or generate it.
+    *
+    * @param tpe General type to be pickled
+    * @param concreteTpe Concrete type that caused the error
+    * @param cause Exception causing the pickling exception if any
+    */
+  final case class FailedPicklerGeneration(tpe: String, concreteTpe: String,
+    override val cause: Option[Throwable] = None) extends BasePicklingException(
+      failedGenerationMsg(tpe, concreteTpe, "pickler"), cause
+  )
+
+  /** Exception thrown when the generation of a [[Unpickler]] for a type
+    * fails because scala pickling wasn't able to find or generate it.
+    *
+    * @param tpe General type to be pickled
+    * @param concreteTpe Concrete type that caused the error
+    * @param cause Exception causing the pickling exception if any
+    */
+  final case class FailedUnpicklerGeneration(tpe: String, concreteTpe: String,
+    override val cause: Option[Throwable] = None) extends BasePicklingException(
+      failedGenerationMsg(tpe, concreteTpe, "unpickler"), cause
+  )
+
+  /** Exception thrown when it's not possible to generated a [[Pickler]]
+    * or [[Unpickler]] at runtime because runtime generation has been disabled.*/
+  final case class RuntimeGenerationDisabled(tpe: String, failed: String)
+    extends UnsupportedOperationException(
+      disabledRuntimeGenerationMsg(tpe, failed)
+    )
+
+  /** Represent any error related to parsing. */
+  class ParsingException(msg: String) extends BasePicklingException(msg)
+
+  /** Exception thrown when the parsing of a message is not successful.
+    *
+    * @param value Message that gave the error
+    */
+  final case class JsonParseFailed(value: String)
```

Comment:
This should probably be in the `json` library, similar for binary. 

Theoretically the json + binary formats are "standalone libs" (or could be).